lib/breadcrumbs.ts

```diff
@@ -0,0 +1,182 @@
+import type { AsyncLocalStorage } from "async_hooks";
+import type { Breadcrumb, InternalBreadcrumb } from "./types";
+import type { Request, Response } from "express";
+const debug = require("debug")("raygun").bind(null, "[breadcrumbs]");
+
+let asyncLocalStorage: AsyncLocalStorage<InternalBreadcrumb[]> | null = null;
+
+try {
+  asyncLocalStorage = new (require("async_hooks").AsyncLocalStorage)();
+  debug("initialized successfully");
+} catch (e) {
+  debug(
+    "failed to load async_hooks.AsyncLocalStorage - initialization failed\n",
+    e,
+  );
+}
+
+type SourceFile = {
+  fileName: string;
+  functionName: string;
+  lineNumber: number | null;
+};
+
+function returnCallerSite(
+  err: Error,
+  callsites: NodeJS.CallSite[],
+): SourceFile | null {
+  for (const callsite of callsites) {
+    const fileName = callsite.getFileName() || "";
+
+    if (fileName.startsWith(__dirname)) {
+      continue;
+    }
+
+    return {
+      fileName: fileName,
+      functionName: callsite.getFunctionName() || "<anonymous>",
+      lineNumber: callsite.getLineNumber(),
+    };
+  }
+
+  return null;
+}
+
+function getCallsite(): SourceFile | null {
+  const originalPrepareStacktrace = Error.prepareStackTrace;
+
+  Error.prepareStackTrace = returnCallerSite;
+
+  // Ignore use of any, required for captureStackTrace
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const output: any = {};
+
+  Error.captureStackTrace(output);
+
+  const callsite = output.stack;
+  Error.prepareStackTrace = originalPrepareStacktrace;
+  return callsite;
+}
+
+export function addBreadcrumb(
+  breadcrumb: string | Breadcrumb,
+  type: InternalBreadcrumb["type"] = "manual",
+) {
+  const crumbs = getBreadcrumbs();
+
+  debug("using breadcrum store:", crumbs);
+
+  if (!crumbs) {
+    return;
+  }
+
+  if (typeof breadcrumb === "string") {
+    const expandedBreadcrumb: Breadcrumb = {
+      message: breadcrumb,
+      level: "info",
+      category: "",
+    };
+
+    breadcrumb = expandedBreadcrumb;
+  }
+
+  const callsite = getCallsite();
+
+  const internalCrumb: InternalBreadcrumb = {
+    ...(breadcrumb as Breadcrumb),
+    category: breadcrumb.category || "",
+    message: breadcrumb.message || "",
+    level: breadcrumb.level || "info",
+    timestamp: Number(new Date()),
+    type,
+    className: callsite?.fileName,
+    methodName: callsite?.functionName,
+
+    // TODO - do we need to do any source mapping?
+    lineNumber: callsite?.lineNumber || undefined,
```

Comment:
Comment was in the original implementation of `breadcrumbs.ts`

Reply:
@miquelbeltran Maybe open an issue on this?

Reply:
Done https://github.com/MindscapeHQ/raygun4node/issues/224